### PR TITLE
Allow custom metadata for replay file

### LINKED ISSUE #78: Allow Custom Data in Replay Files
## **Description**  

Hey folks! 👋  

I have a use case where I need to add extra data to the replay file so I can **properly reproduce my tests**. One specific example is when a test involves **random number generation**—I need to store the **seed** that was used. Since each test might have a different seed, this would be super useful to ensure reproducibility.  

## **Proposal**  

It would be great if ``pytest-replay`` allowed users to **store custom data** in the replay file. This could be done via a fixture or some other mechanism that lets users define extra metadata when a test runs. Then, when replaying, this data would be available to ensure the test environment is correctly reconstructed.  

## **Implementation Idea**  

- Introduce a way to let users **attach custom metadata** (maybe via fixture or another hook, or custom hook? Maybe a way to store some fixture values?).
- Store this data alongside the standard replay info.  
- When replaying, load the metadata so tests can use it.  
- This could be useful for things beyond random seeds too!  

## **Limitations**  

I know there are **some edge cases**, like when a **core dump** happens before the data is written. But even with that limitation, I think this would be a really nice addition for a lot of users (at least for me! 😄).  

## **Happy to Contribute** 🚀  

If this sounds like a useful feature, I’d be happy to create a PR and implement it! Let me know what you think. 😊  

---

cc: @tadeu @nicoddemus @prusse-martin

### PR DESCRIPTION
**Pull request by marcelotrevisani**
Allow custom metadata in replay files
    
Refactor replay metadata handling with dataclasses. Introduced `ReplayTestMetadata` dataclass for cleaner metadata management. Updated serialization and deserialization logic to use the new dataclass. Added new tests to validate metadata handling and functionality.

Closes #78

### CODE REVIEW

**Comment by marcelotrevisani** [comment]
that is ready for review please
@nicoddemus @tadeu @prusse-martin

File: README.rst [hunk 1]
```diff
@@ -69,6 +69,38 @@ execute the tests in the same order with::
 
 Hopefully this will make it easier to reproduce the problem and fix it.
 
+Additional metadata
+-------------------
+
```
Comment (nicoddemus) [inline README.rst:74]: Great docs!

```suggestion

*Version added: 1.6*

```

File: src/pytest_replay/__init__.py [hunk 2]
```diff
@@ -39,6 +44,24 @@ def pytest_addoption(parser):
     )
 
 
+@dataclasses.dataclass
+class ReplayTestMetadata:
```
Comment (nicoddemus) [inline src/pytest_replay/__init__.py:48]: This class is not only responsible for the metadata, but everything else related to the plugin, like start, outcome, etc. To avoid confusing it with the "metadata" attribute, how about naming it differently, say:

```suggestion
class ReplayTestInfo:
```

File: src/pytest_replay/__init__.py [hunk 3]
```diff
@@ -39,6 +44,24 @@ def pytest_addoption(parser):
...
+    start: float = 0.0
+    finish: Optional[float] = None
+    outcome: Optional[str] = None
+    metadata: dict[str, Any] = dataclasses.field(default_factory=dict)
+
+    def to_clean_dict(self) -> dict[str, Any]:
+        return {k: v for k, v in asdict(self).items() if v}
+
+
+class _ReplayTestMetadataDefaultDict(collections.defaultdict):
```
Comment (nicoddemus) [inline src/pytest_replay/__init__.py:59]: As per the other comment, for consistency:

```suggestion
class _ReplayTestInfoDefaultDict(collections.defaultdict):
```

File: src/pytest_replay/__init__.py [hunk 4]
```diff
@@ -119,7 +142,10 @@ def pytest_collection_modifyitems(self, items, config):
                 stripped = line.strip()
                 # Ignore blank linkes and comments. (#70)
                 if stripped and not stripped.startswith(("#", "//")):
-                    nodeid = json.loads(stripped)["nodeid"]
+                    node_metadata = json.loads(stripped)
+                    nodeid = node_metadata["nodeid"]
+                    if "finish" in node_metadata:
+                        self.nodes[nodeid] = ReplayTestMetadata(**node_metadata)
```
Comment (nicoddemus) [inline src/pytest_replay/__init__.py:148]:
```suggestion
                    node_info = json.loads(stripped)
                    nodeid = node_info["nodeid"]
                    if "finish" in node_info:
                        self.nodes[nodeid] = ReplayTestInfo(**node_info)
```

File: tests/test_replay.py [hunk 5]
```diff
@@ -274,6 +274,68 @@ def test_filter_out_tests_not_in_file(testdir):
...
+    pytester.maketxtfile(
+        """{\"nodeid\": \"test_replay_extra_metadata_load.py::test_load\", \"start\": 1.0}
+    {\"nodeid\": \"test_replay_extra_metadata_load.py::test_load\", \"start\": 1.0, \"finish\": 2.0, \"outcome\": \"passed\", \"metadata\": {\"seed\": 1234}}
+    """
+    )
+    result = pytester.runpytest(
+        f"--replay={pytester.path / 'test_replay_extra_metadata_load.txt'}"
+    )
+    assert result.ret == 0
+    result.assert_outcomes(passed=1)
```
Comment (nicoddemus) [inline tests/test_replay.py:336]: I think we can unify those two tests into one:

```suggestion
def test_metadata(pytester, tmp_path):
    pytester.makepyfile(
        """
        import pytest

        @pytest.fixture
        def seed(replay_metadata):
            assert replay_metadata.metadata == {}
            replay_metadata.metadata["seed"] = seed = 1234
            return seed
        
        def test_foo(seed):
            assert seed == 1234
        """
    )
    dir = tmp_path / "replay"
    result = pytester.runpytest(f"--replay-record-dir={dir}")
    assert result.ret == 0

    # Rewrite the fixture to always returns the metadata, as written previously.
    pytester.makepyfile(
        """
        import pytest

        @pytest.fixture
        def seed(replay_metadata):
            return replay_metadata.metadata["seed"]            
        
        def test_foo(seed):
            assert seed == 1234
        """
    )
    result = pytester.runpytest(f"--replay={dir / '.pytest-replay.txt'}")
    assert result.ret == 0
```

This is simpler, simulates the actual workflow, and avoids having to write the replay data manually. 👍

**Review by nicoddemus — CHANGES_REQUESTED** [review]
Great work!

Left some comments, please take a look.

File: src/pytest_replay/__init__.py (hunk 4 above)
Comment (prusse-martin) [inline src/pytest_replay/__init__.py:148]: I believe that if somehow the test crashes before the finish you metadata would no be save, but it could be relevant to the crash reproduction.

You possible want to emit to the replay file "update metadata" instructions so they could be loaded back to reproduce the problem.
Something in the lines of:
```json
{"nodeid": "test_1.py::test_bar", "start": 3.0}
{"nodeid": "test_1.py::test_bar", "start": 3.0, "metadata": {"a": "asd"}}
{"nodeid": "test_1.py::test_bar", "start": 3.0, "metadata": {"a": "asd", "b": 7}}
{"nodeid": "test_1.py::test_bar", "start": 3.0, "finish": 4.0, "outcome": "passed"}
```
There we see the metadata evolving.

And probably could help pinpoint the crash source if the output is just:
```json
{"nodeid": "test_1.py::test_bar", "start": 3.0}
{"nodeid": "test_1.py::test_bar", "start": 3.0, "metadata": {"a": "asd"}}
```
(a crash cause some lines to not be present)

File: src/pytest_replay/__init__.py (hunk 2 above)
Comment (marcelotrevisani) [reply src/pytest_replay/__init__.py:48]: I agree, ``ReplayTestInfo`` is better, indeed

File: src/pytest_replay/__init__.py (hunk 4 above)
Comment (marcelotrevisani) [reply src/pytest_replay/__init__.py]: Thanks for the feedback! You're absolutely right, and this is something I already noted in the issue under "Limitations". In my specific scenario, it's quite rare for tests to crash, they usually just fail, so the current approach works well enough for my use case. However, I do see the value in capturing incremental metadata.
To fully address this broader scenario, we'd likely need to track all setups and all teardowns associated with a test, ensuring the metadata reflects each setup and teardown, since a crash could happen at any point.

For now, I’d prefer to move forward with the current implementation, but I completely agree that your suggestion would be the ideal enhancement. It’s definitely something worth considering for future improvements! Maybe, could you create an issue about it so we can tackle that in the future or when someone sees values in it please?

**Comment by marcelotrevisani** [comment]
> Great work!
> 
> Left some comments, please take a look.

Thank you for your feedback!

I have addressed all your comments in the review, could you please take a second look? Thanks!

**Review by nicoddemus — APPROVED** [review]
Nice, thanks!

After @prusse-martin's approval we can make a new release.
